File: java/src/main/java/io/yosina/transliterators/HistoricalHirakatasTransliterator.java
package io.yosina.transliterators;

import io.yosina.Char;
import io.yosina.CharIterator;
import io.yosina.CodePointTuple;
import io.yosina.Transliterator;
import io.yosina.annotations.RegisteredTransliterator;
import java.util.NoSuchElementException;
import java.util.Objects;

/**
 * Historical hiragana/katakana transliterator.
 *
 * <p>This transliterator converts historical hiragana and katakana characters to their modern
 * equivalents. It supports three conversion modes for each character category:
 *
 * <ul>
 *   <li>SIMPLE: replaces historical characters with their closest modern equivalents
 *   <li>DECOMPOSE: replaces historical characters with decomposed modern equivalents
 *   <li>SKIP: leaves historical characters unchanged
 * </ul>
 *
 * <p>Historical hiragana mappings:
 *
 * <ul>
 *   <li>{@code ゐ} (U+3090) -> {@code い} (simple) / {@code うぃ} (decompose)
 *   <li>{@code ゑ} (U+3091) -> {@code え} (simple) / {@code うぇ} (decompose)
 * </ul>
 *
 * <p>Historical katakana mappings:
 *
 * <ul>
 *   <li>{@code ヰ} (U+30F0) -> {@code イ} (simple) / {@code ウィ} (decompose)
 *   <li>{@code ヱ} (U+30F1) -> {@code エ} (simple) / {@code ウェ} (decompose)
 * </ul>
 *
 * <p>Voiced historical kana mappings (decompose only):
 *
 * <ul>
 *   <li>{@code ヷ} (U+30F7) -> {@code ヴァ}
 *   <li>{@code ヸ} (U+30F8) -> {@code ヴィ}
 *   <li>{@code ヹ} (U+30F9) -> {@code ヴェ}
 *   <li>{@code ヺ} (U+30FA) -> {@code ヴォ}
 * </ul>
 */
@RegisteredTransliterator(name = "historical-hirakatas")
public class HistoricalHirakatasTransliterator implements Transliterator {

    /** Conversion mode for historical kana characters. */
    public enum ConversionMode {
        /** Replace with the closest modern equivalent. */
        SIMPLE,
        /** Replace with decomposed modern equivalent. */
        DECOMPOSE,
        /** Leave historical characters unchanged. */
        SKIP
    }

    /** Options for configuring the behavior of HistoricalHirakatasTransliterator. */
    public static class Options {
        private final ConversionMode hiraganas;
        private final ConversionMode katakanas;
        private final ConversionMode voicedKatakanas;

        /** Creates a new Options instance with default settings. */
        public Options() {
            this(ConversionMode.SIMPLE, ConversionMode.SIMPLE, ConversionMode.SKIP);
        }

        /**
         * Creates a new Options instance with the specified settings.
         *
         * @param hiraganas the conversion mode for historical hiragana
         * @param katakanas the conversion mode for historical katakana
         * @param voicedKatakanas the conversion mode for voiced historical kana
         */
        public Options(
                ConversionMode hiraganas,
                ConversionMode katakanas,
                ConversionMode voicedKatakanas) {
            this.hiraganas = hiraganas;
            this.katakanas = katakanas;
            this.voicedKatakanas = voicedKatakanas;
        }

        /**
         * Returns the conversion mode for historical hiragana.
         *
         * @return the conversion mode for historical hiragana
         */
        public ConversionMode getHiraganas() {
            return hiraganas;
        }

        /**
         * Returns the conversion mode for historical katakana.
         *
         * @return the conversion mode for historical katakana
         */
        public ConversionMode getKatakanas() {
            return katakanas;
        }

        /**
         * Returns the conversion mode for voiced historical kana.
         *
         * @return the conversion mode for voiced historical kana
         */
        public ConversionMode getVoicedKatakanas() {
            return voicedKatakanas;
        }

        /**
         * Creates a new Options instance with the specified historical hiragana conversion mode.
         *
         * @param mode the conversion mode for historical hiragana
         * @return a new Options instance with the updated setting
         */
        public Options withHiraganas(ConversionMode mode) {
            return new Options(mode, this.katakanas, this.voicedKatakanas);
        }

        /**
         * Creates a new Options instance with the specified historical katakana conversion mode.
         *
         * @param mode the conversion mode for historical katakana
         * @return a new Options instance with the updated setting
         */
        public Options withKatakanas(ConversionMode mode) {
            return new Options(this.hiraganas, mode, this.voicedKatakanas);
        }

        /**
         * Creates a new Options instance with the specified voiced historical kana conversion mode.
         *
         * @param mode the conversion mode for voiced historical kana
         * @return a new Options instance with the updated setting
         */
        public Options withVoicedKatakanas(ConversionMode mode) {
            return new Options(this.hiraganas, this.katakanas, mode);
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            Options options = (Options) obj;
            return hiraganas == options.hiraganas
                    && katakanas == options.katakanas
                    && voicedKatakanas == options.voicedKatakanas;
        }

        @Override
        public int hashCode() {
            return Objects.hash(hiraganas, katakanas, voicedKatakanas);
        }
    }

    // Historical hiragana code points
    private static final int WI_HIRAGANA = 0x3090; // ゐ
    private static final int WE_HIRAGANA = 0x3091; // ゑ

    // Historical katakana code points
    private static final int WI_KATAKANA = 0x30F0; // ヰ
    private static final int WE_KATAKANA = 0x30F1; // ヱ

    // Base katakana that compose with combining dakuten to form voiced historical kana
    private static final int WA_KATAKANA = 0x30EF; // ワ
    private static final int WO_KATAKANA = 0x30F2; // ヲ
    private static final int COMBINING_DAKUTEN = 0x3099; // ゙

    // Voiced historical katakana code points
    private static final int VA_KATAKANA = 0x30F7; // ヷ
    private static final int VI_KATAKANA = 0x30F8; // ヸ
    private static final int VE_KATAKANA = 0x30F9; // ヹ
    private static final int VO_KATAKANA = 0x30FA; // ヺ

    // Simple replacements
    private static final CodePointTuple SIMPLE_I_HIRAGANA = CodePointTuple.of(0x3044); // い
    private static final CodePointTuple SIMPLE_E_HIRAGANA = CodePointTuple.of(0x3048); // え
    private static final CodePointTuple SIMPLE_I_KATAKANA = CodePointTuple.of(0x30A4); // イ
    private static final CodePointTuple SIMPLE_E_KATAKANA = CodePointTuple.of(0x30A8); // エ

    // Decomposed replacements
    private static final CodePointTuple DECOMPOSE_WI_HIRAGANA =
            CodePointTuple.of(0x3046, 0x3043); // うぃ
    private static final CodePointTuple DECOMPOSE_WE_HIRAGANA =
            CodePointTuple.of(0x3046, 0x3047); // うぇ
    private static final CodePointTuple DECOMPOSE_WI_KATAKANA =
            CodePointTuple.of(0x30A6, 0x30A3); // ウィ
    private static final CodePointTuple DECOMPOSE_WE_KATAKANA =
            CodePointTuple.of(0x30A6, 0x30A7); // ウェ

    // Voiced historical replacements (decompose only, for composed input)
    // VU prefix is emitted separately; these are just the small vowel suffixes.
    private static final int VU_KATAKANA = 0x30F4; // ヴ
    private static final int U_KATAKANA = 0x30A6; // ウ
    private static final CodePointTuple VU_TUPLE = CodePointTuple.of(VU_KATAKANA);
    private static final CodePointTuple U_TUPLE = CodePointTuple.of(U_KATAKANA);

    private static int getVoicedVowel(int composedCodePoint) {
        switch (composedCodePoint) {
            case VA_KATAKANA:
                return 0x30A1; // ァ
            case VI_KATAKANA:
                return 0x30A3; // ィ
            case VE_KATAKANA:
                return 0x30A7; // ェ
            case VO_KATAKANA:
                return 0x30A9; // ォ
            default:
                return -1;
        }
    }

    private static int getDecomposedVowel(int base) {
        switch (base) {
            case WA_KATAKANA:
                return 0x30A1; // ァ
            case WI_KATAKANA:
                return 0x30A3; // ィ
            case WE_KATAKANA:
                return 0x30A7; // ェ
            case WO_KATAKANA:
                return 0x30A9; // ォ
            default:
                return -1;
        }
    }

    private final Options options;

    /** Creates a new HistoricalHirakatasTransliterator with default options. */
    public HistoricalHirakatasTransliterator() {
        this(new Options());
    }

    /**
     * Creates a new HistoricalHirakatasTransliterator with the specified options.
     *
     * @param options the options for configuring the transliterator behavior
     */
    public HistoricalHirakatasTransliterator(Options options) {
        this.options = options;
    }

    /**
     * Transliterates the input by converting historical hiragana/katakana characters.
     *
     * @param input the input character iterator
     * @return a new character iterator with converted characters
     */
    @Override
    public CharIterator transliterate(CharIterator input) {
        return new HistoricalHirakatasCharIterator(input, options);
    }

    private static class HistoricalHirakatasCharIterator implements CharIterator {
        private final CharIterator input;
        private final Options options;
        private int offset = 0;
        private Char pending = null;
        private final java.util.ArrayDeque<Char> queue = new java.util.ArrayDeque<>();

        public HistoricalHirakatasCharIterator(CharIterator input, Options options) {
            this.input = input;
            this.options = options;
        }

        @Override
        public boolean hasNext() {
            return !queue.isEmpty() || pending != null || input.hasNext();
        }

        @Override
        public Char next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }

            // Drain queued chars first
            if (!queue.isEmpty()) {
                return queue.poll();
            }

            final Char c;
            if (pending != null) {
                c = pending;
                pending = null;
            } else {
                c = input.next();
            }
            if (c == null || c.isSentinel()) {
                return c;
            }

            final CodePointTuple cpt = c.get();
            if (cpt.size() != 1) {
                Char result = c.withOffset(offset);
                offset += result.charCount();
                return result;
            }

            final int codePoint = cpt.get(0);

            // Lookahead: peek at next char for combining dakuten
            if (input.hasNext()) {
                Char nextChar = input.next();
                if (nextChar != null
                        && !nextChar.isSentinel()
                        && nextChar.get().size() == 1
                        && nextChar.get().get(0) == COMBINING_DAKUTEN) {
                    // Check if current char is a decomposed voiced base
                    int vowel = getDecomposedVowel(codePoint);
                    if (vowel != -1 && options.voicedKatakanas == ConversionMode.DECOMPOSE) {
                        // decompose mode: emit U + dakuten + vowel
                        Char resultU = new Char(U_TUPLE, offset, c);
                        offset += U_TUPLE.charCount();
                        Char dakutenChar = nextChar.withOffset(offset);
                        offset += dakutenChar.charCount();
                        queue.add(dakutenChar);
                        CodePointTuple vowelTuple = CodePointTuple.of(vowel);
                        Char vowelChar = new Char(vowelTuple, offset, c);
                        offset += vowelTuple.charCount();
                        queue.add(vowelChar);
                        return resultU;
                    } else {
                        // skip mode or not a voiced base: store dakuten in pending, return base
                        // unchanged
                        pending = nextChar;
                        Char result = c.withOffset(offset);
                        offset += result.charCount();
                        return result;
                    }
                } else {
                    // Next char is not dakuten, store it in pending
                    pending = nextChar;
                }
            }

            final CodePointTuple replacement = getReplacement(codePoint);

            if (replacement != null) {
                Char result = new Char(replacement, offset, c);
                offset += replacement.charCount();
                return result;
            }

            Char result = c.withOffset(offset);
            offset += result.charCount();
            return result;
        }

        private CodePointTuple getReplacement(int codePoint) {
            switch (codePoint) {
                // Historical hiragana
                case WI_HIRAGANA:
                    return getHistoricalHiraganaReplacement(
                            SIMPLE_I_HIRAGANA, DECOMPOSE_WI_HIRAGANA);
                case WE_HIRAGANA:
                    return getHistoricalHiraganaReplacement(
                            SIMPLE_E_HIRAGANA, DECOMPOSE_WE_HIRAGANA);

                // Historical katakana
                case WI_KATAKANA:
                    return getHistoricalKatakanaReplacement(
                            SIMPLE_I_KATAKANA, DECOMPOSE_WI_KATAKANA);
                case WE_KATAKANA:
                    return getHistoricalKatakanaReplacement(
                            SIMPLE_E_KATAKANA, DECOMPOSE_WE_KATAKANA);

                // Voiced historical katakana
                case VA_KATAKANA:
                case VI_KATAKANA:
                case VE_KATAKANA:
                case VO_KATAKANA:
                    return getVoicedHistoricalReplacement(codePoint);

                default:
                    return null;
            }
        }

        private CodePointTuple getHistoricalHiraganaReplacement(
                CodePointTuple simple, CodePointTuple decompose) {
            switch (options.hiraganas) {
                case SIMPLE:
                    return simple;
                case DECOMPOSE:
                    return decompose;
                case SKIP:
                default:
                    return null;
            }
        }

        private CodePointTuple getHistoricalKatakanaReplacement(
                CodePointTuple simple, CodePointTuple decompose) {
            switch (options.katakanas) {
                case SIMPLE:
                    return simple;
                case DECOMPOSE:
                    return decompose;
                case SKIP:
                default:
                    return null;
            }
        }

        private CodePointTuple getVoicedHistoricalReplacement(int codePoint) {
            if (options.voicedKatakanas != ConversionMode.DECOMPOSE) {
                return null;
            }
            int vowel = getVoicedVowel(codePoint);
            if (vowel == -1) {
                return null;
            }
            return CodePointTuple.of(VU_KATAKANA, vowel);
        }
    }
}

File: csharp/src/Yosina/Transliterators/HistoricalHirakatasTransliterator.cs
// Copyright (c) Yosina. All rights reserved.

using System.Text.Json.Serialization;
using Yosina.JsonConverters;

namespace Yosina.Transliterators;

/// <summary>
/// Transliterator for converting historical hiragana/katakana characters to their modern equivalents.
/// </summary>
[RegisteredTransliterator("historical-hirakatas")]
public class HistoricalHirakatasTransliterator : ITransliterator
{
    /// <summary>
    /// Conversion mode for historical kana characters.
    /// </summary>
    [JsonConverter(typeof(JsonEnumValueConverter<ConversionMode>))]
    public enum ConversionMode
    {
        /// <summary>
        /// Replace with the simple modern equivalent.
        /// </summary>
        [JsonEnumValue("simple")]
        Simple,

        /// <summary>
        /// Decompose into the phonetic equivalent using multiple modern characters.
        /// </summary>
        [JsonEnumValue("decompose")]
        Decompose,

        /// <summary>
        /// Skip conversion; leave the character unchanged.
        /// </summary>
        [JsonEnumValue("skip")]
        Skip,
    }

    /// <summary>
    /// Conversion mode for voiced historical kana characters.
    /// </summary>
    [JsonConverter(typeof(JsonEnumValueConverter<VoicedConversionMode>))]
    public enum VoicedConversionMode
    {
        /// <summary>
        /// Decompose into the phonetic equivalent using multiple modern characters.
        /// </summary>
        [JsonEnumValue("decompose")]
        Decompose,

        /// <summary>
        /// Skip conversion; leave the character unchanged.
        /// </summary>
        [JsonEnumValue("skip")]
        Skip,
    }

    /// <summary>
    /// Options for configuring the behavior of HistoricalHirakatasTransliterator.
    /// </summary>
    public class Options
    {
        /// <summary>
        /// Gets or sets the conversion mode for historical hiragana characters (e.g. ゐ, ゑ).
        /// </summary>
        [JsonPropertyName("hiraganas")]
        public ConversionMode Hiraganas { get; set; } = ConversionMode.Simple;

        /// <summary>
        /// Gets or sets the conversion mode for historical katakana characters (e.g. ヰ, ヱ).
        /// </summary>
        [JsonPropertyName("katakanas")]
        public ConversionMode Katakanas { get; set; } = ConversionMode.Simple;

        /// <summary>
        /// Gets or sets the conversion mode for voiced historical kana characters (e.g. ヷ, ヸ, ヹ, ヺ).
        /// </summary>
        [JsonPropertyName("voicedKatakanas")]
        public VoicedConversionMode VoicedKatakanas { get; set; } = VoicedConversionMode.Skip;
    }

    private const int CombiningDakuten = 0x3099;
    private const int Vu = 0x30F4; // ヴ
    private const int U = 0x30A6; // ウ

    private static readonly Dictionary<int, int> VoicedDecomposedMappings = new()
    {
        [0x30EF] = 0x30A1, // ヷ → ァ
        [0x30F0] = 0x30A3, // ヸ → ィ
        [0x30F1] = 0x30A7, // ヹ → ェ
        [0x30F2] = 0x30A9, // ヺ → ォ
    };

    private readonly Dictionary<int, int[]> mappings;
    private readonly VoicedConversionMode voicedKatakanas;

    /// <summary>
    /// Initializes a new instance of the <see cref="HistoricalHirakatasTransliterator"/> class with the specified options.
    /// </summary>
    /// <param name="options">The options for the transliterator.</param>
    public HistoricalHirakatasTransliterator(Options options)
    {
        this.mappings = BuildMappings(options);
        this.voicedKatakanas = options.VoicedKatakanas;
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="HistoricalHirakatasTransliterator"/> class with default options.
    /// </summary>
    public HistoricalHirakatasTransliterator()
        : this(new Options())
    {
    }

    private static Dictionary<int, int[]> BuildMappings(Options options)
    {
        var result = new Dictionary<int, int[]>();

        // Historical hiragana mappings
        switch (options.Hiraganas)
        {
            case ConversionMode.Simple:
                result[0x3090] = new[] { 0x3044 }; // ゐ (U+3090) -> い (U+3044)
                result[0x3091] = new[] { 0x3048 }; // ゑ (U+3091) -> え (U+3048)
                break;
            case ConversionMode.Decompose:
                result[0x3090] = new[] { 0x3046, 0x3043 }; // ゐ (U+3090) -> うぃ (U+3046, U+3043)
                result[0x3091] = new[] { 0x3046, 0x3047 }; // ゑ (U+3091) -> うぇ (U+3046, U+3047)
                break;
            case ConversionMode.Skip:
                break;
        }

        // Historical katakana mappings
        switch (options.Katakanas)
        {
            case ConversionMode.Simple:
                result[0x30F0] = new[] { 0x30A4 }; // ヰ (U+30F0) -> イ (U+30A4)
                result[0x30F1] = new[] { 0x30A8 }; // ヱ (U+30F1) -> エ (U+30A8)
                break;
            case ConversionMode.Decompose:
                result[0x30F0] = new[] { 0x30A6, 0x30A3 }; // ヰ (U+30F0) -> ウィ (U+30A6, U+30A3)
                result[0x30F1] = new[] { 0x30A6, 0x30A7 }; // ヱ (U+30F1) -> ウェ (U+30A6, U+30A7)
                break;
            case ConversionMode.Skip:
                break;
        }

        // Voiced historical kana mappings
        switch (options.VoicedKatakanas)
        {
            case VoicedConversionMode.Decompose:
                result[0x30F7] = new[] { Vu, 0x30A1 }; // ヷ (U+30F7) -> ヴァ
                result[0x30F8] = new[] { Vu, 0x30A3 }; // ヸ (U+30F8) -> ヴィ
                result[0x30F9] = new[] { Vu, 0x30A7 }; // ヹ (U+30F9) -> ヴェ
                result[0x30FA] = new[] { Vu, 0x30A9 }; // ヺ (U+30FA) -> ヴォ
                break;
            case VoicedConversionMode.Skip:
                break;
        }

        return result;
    }

    /// <inheritdoc/>
    public IEnumerable<Character> Transliterate(IEnumerable<Character> input)
    {
        return new HistoricalHirakatasEnumerator(input.GetEnumerator(), this.mappings, this.voicedKatakanas);
    }

    private class HistoricalHirakatasEnumerator : IEnumerable<Character>, IEnumerator<Character>
    {
        private readonly IEnumerator<Character> input;
        private readonly Dictionary<int, int[]> mappings;
        private readonly VoicedConversionMode voicedKatakanas;
        private int offset;
        private bool disposed;
        private readonly Queue<Character> queue = new();
        private Character? pending;

        public HistoricalHirakatasEnumerator(IEnumerator<Character> input, Dictionary<int, int[]> mappings, VoicedConversionMode voicedKatakanas)
        {
            this.input = input;
            this.mappings = mappings;
            this.voicedKatakanas = voicedKatakanas;
        }

        public Character Current { get; private set; } = null!;

        object System.Collections.IEnumerator.Current => this.Current;

        public bool MoveNext()
        {
            if (this.disposed)
            {
                return false;
            }

            // If we have queued characters from a previous expansion, return them first
            if (this.queue.Count > 0)
            {
                this.Current = this.queue.Dequeue();
                return true;
            }

            Character c;
            if (this.pending != null)
            {
                c = this.pending;
                this.pending = null;
            }
            else if (this.input.MoveNext())
            {
                c = this.input.Current;
            }
            else
            {
                return false;
            }

            if (c.IsSentinel)
            {
                this.Current = c.WithOffset(this.offset);
                return true;
            }

            // Lookahead: peek at next char for combining dakuten
            if (c.CodePoint.Size() == 1)
            {
                bool hasNext;
                Character? next = null;
                if (this.input.MoveNext())
                {
                    hasNext = true;
                    next = this.input.Current;
                }
                else
                {
                    hasNext = false;
                }

                if (hasNext && next != null && !next.IsSentinel && next.CodePoint.Size() == 1 && next.CodePoint.First == CombiningDakuten)
                {
                    // Check if current char is a decomposed voiced base
                    if (this.voicedKatakanas == VoicedConversionMode.Decompose && VoicedDecomposedMappings.TryGetValue(c.CodePoint.First, out var vowel))
                    {
                        // Decompose mode: emit U + dakuten + vowel
                        this.Current = new Character((U, -1), this.offset, c);
                        this.offset += this.Current.CharCount;

                        var dakutenChar = next.WithOffset(this.offset);
                        this.queue.Enqueue(dakutenChar);
                        this.offset += dakutenChar.CharCount;

                        var vowelChar = new Character((vowel, -1), this.offset, c);
                        this.queue.Enqueue(vowelChar);
                        this.offset += vowelChar.CharCount;
                    }
                    else
                    {
                        // Skip mode or not a voiced base: emit current char, then queue the dakuten char
                        this.Current = c.WithOffset(this.offset);
                        this.offset += this.Current.CharCount;
                        var dakutenChar = next.WithOffset(this.offset);
                        this.queue.Enqueue(dakutenChar);
                        this.offset += dakutenChar.CharCount;
                    }
                }
                else
                {
                    // No dakuten follows: store next as pending and fall through to regular mapping
                    if (hasNext)
                    {
                        this.pending = next;
                    }

                    if (this.mappings.TryGetValue(c.CodePoint.First, out var mapped))
                    {
                        if (mapped.Length > 0)
                        {
                            this.Current = new Character((mapped[0], -1), this.offset, c);
                            this.offset += this.Current.CharCount;

                            for (int i = 1; i < mapped.Length; i++)
                            {
                                var ch = new Character((mapped[i], -1), this.offset, c);
                                this.queue.Enqueue(ch);
                                this.offset += ch.CharCount;
                            }
                        }
                        else
                        {
                            this.Current = c.WithOffset(this.offset);
                            this.offset += this.Current.CharCount;
                        }
                    }
                    else
                    {
                        this.Current = c.WithOffset(this.offset);
                        this.offset += this.Current.CharCount;
                    }
                }
            }
            else
            {
                this.Current = c.WithOffset(this.offset);
                this.offset += this.Current.CharCount;
            }

            return true;
        }

        public void Reset()
        {
            this.input.Reset();
            this.offset = 0;
            this.queue.Clear();
            this.pending = null;
        }

        public void Dispose()
        {
            if (!this.disposed)
            {
                this.input.Dispose();
                this.disposed = true;
            }
        }

        public IEnumerator<Character> GetEnumerator() => this;

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => this;
    }
}
